pkg/certificate/signing_requestor.go

```diff
@@ -0,0 +1,487 @@
+		s.brokerSecretClient.DeleteCollection(ctx, metav1.DeleteOptions{}, listOpts))
+}
+
+func (s *signingRequestorImpl) SetKeyGenerator(kg KeyGeneratorFn) {
+	s.keyGenerator = kg
+}
+
+func (s *signingRequestorImpl) generateKeyAndCSR(ips []string) ([]byte, []byte, error) {
+	privateKey, err := rsa.GenerateKey(rand.Reader, RSABitSize)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ipAddresses := []net.IP{}
+	for _, ip := range ips {
+		parsed := net.ParseIP(ip)
+		if parsed == nil {
+			return nil, nil, fmt.Errorf("invalid IP address in SAN: %s", ip)
+		}
+		ipAddresses = append(ipAddresses, parsed)
+	}
+
+	csrTemplate := x509.CertificateRequest{
+		Subject: pkix.Name{
+			CommonName:   fmt.Sprintf("submariner-%s", s.localClusterID),
+			Organization: []string{"submariner.io"},
+		},
+		SignatureAlgorithm: x509.SHA256WithRSA,
+		IPAddresses:        ipAddresses,
+	}
+
+	csrDER, err := x509.CreateCertificateRequest(rand.Reader, &csrTemplate, privateKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	keyPEM := pem.EncodeToMemory(&pem.Block{Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(privateKey)})
+	csrPEM := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE REQUEST", Bytes: csrDER})
+
+	return keyPEM, csrPEM, nil
+}
+
+func (s *signingRequestorImpl) onLocalSecretSigned(obj runtime.Object, numRequeues int) bool {
+	// warningLogInterval defines how often to log warnings for missing callbacks.
+	const warningLogInterval = 12
+
+	secret := obj.(*corev1.Secret)
+
+	v, ok := s.onSignedMap.Load(secret.Name)
+	if !ok {
+		if numRequeues > 0 && numRequeues%warningLogInterval == 0 {
+			logger.Warningf("Received signed Secret for %q with no OnSigned callback registered", secret.Name)
+		}
+
+		return true
+	}
+
+	// Update certificate expiration time for renewal tracking
+	s.updateCertificateExpiration(secret)
+
+	err := v.(OnSignedFn)(secret.Data)
+	if err == nil {
+		return false
+	}
+
+	logger.Errorf(err, "OnSignedFn returned an error for secret %q", secret.Name)
+
+	return true
+}
+
+func (s *signingRequestorImpl) secretName(name string) string {
+	return fmt.Sprintf("%s-%s", name, s.localClusterID)
+}
+
+func deleteIfPresent(ctx context.Context, client dynamic.ResourceInterface, name string) error {
+	err := client.Delete(ctx, name, metav1.DeleteOptions{})
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+
+	return errors.Wrapf(err, "error deleting Secret %q", name)
+}
+
+// startCertificateRenewalMonitoring starts periodic monitoring of certificate expiration
+func (s *signingRequestorImpl) startCertificateRenewalMonitoring(stopCh <-chan struct{}) {
+	s.renewalTicker = time.NewTicker(CertCheckInterval)
+	s.renewalStopCh = make(chan struct{})
+
+	go func() {
+		defer s.renewalTicker.Stop()
+
+		for {
+			select {
+			case <-s.renewalTicker.C:
+				s.checkCertificateRenewal()
+			case <-s.renewalStopCh:
+				return
+			case <-stopCh:
+				return
+			}
+		}
+	}()
```

Comment:
You can use `wait.Until` instead:

```suggestion
	wait.Until(s.checkCertificateRenewal, CertCheckInterval, stopCh)
```